### PR TITLE
Kotlin Gradle scripts. Part 1 of 2

### PR DESCRIPTION
**Pull request by dmdashenkov**
In this PR we start migrating our Gradle scripts to Kotlin.

For simplicity, we split the changes into 2 PRs. In this PR we migrate the basic scripts of the repo's main build. Other builds (`smoke-tests` and `base-validating-builders`) are to be migrated in the next part.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol, @alexander-yevsyukov, PTAL.

File: .travis.yml [hunk 1]
```diff
@@ -11,6 +11,9 @@ env:
     # Encrypted `GCS_SECRET` variable.
     - secure: "i8MhONZu7QjyM2V887A1Tydr1WMqQP5jJZNjIJjc1Uae8F0/z8cJZIZ1hstodN7FpoR4VF92zyhUwbt6fz/dsdPEJFccsiMlEc9vlqecQCd267160wgRZneaB6Xe/y/EUmq9XsGdn/k1Ey+QZwX9au/8RU191v+fDsCtMRYXzyEa/BvbQuSwuYRgQDxTAxuJgTmG5Sxl9jWqKw1BfxUcEoErc/jqymU58w6z2TxKxVzIXT29Jy/Z12VuSiS8opigSrIP8e/1fctC84wI7S52mext2ZfhPYSTHFKS+xg1vQDYPb8m5aomL8E6Of7hVD5BTnEnyjj+/Gr63GAzHXtkHhWoxo+vB+xBFfDu8wxM5Aqna3H7LMDD5kGCxQEz8qmzHBHMAhLnhsRzjNVu2+tLCZdeMN88Ud2uemL2SCAcR8Juleg7DGMj3D0SAbPyUH3+9yYYWzSg6iaxgTdHBnJ+uXUJp0Nu+M2EK6Kl+pYAsCLVfZRPGaajFXVnJEPPeSr2PYzk7F4pIzgn/E8AtYEJ0gcEbjoTItS8EjliJKDXM4HdkluXBFLvzIH1O1nCtxKNv4UkUmPhFbfHrPXDcsYq2zsEe+NkvsJlxjAwYnOMkT4NLiEsec1a7K9bBC+iQA9e8rriMbu6/1w63JErQyx05avPjgO8XRDK8hxTf4rhBmY="
     - GRADLE_OPTS="-Xmx2G"
+    # TODO:2020-05-18:dmytro.dashenkov: Remove there 2 lines when Travis stops failing when installing JDK 8.
```
Comment (alexander-yevsyukov) [inline .travis.yml:14]: `there` -> `these`?

File: base/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    id("com.google.protobuf")
+}
+
+group = "io.spine"
+
+apply(from = Deps.scripts.testArtifacts(project))
+
+configurations {
+    // Avoid collisions of Java classes defined both in `protobuf-lite` and `protobuf-java`
+    runtimeClasspath.get().exclude(group = "com.google.protobuf", module = "protobuf-lite")
```
Comment (alexander-yevsyukov) [inline base/build.gradle.kts:39]: Can we have these constants extracted into final variables?

File: base/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+                }
+            }
+        }
+    })
+}
+
+/**
+ * Checks if the given file belongs to the Google `.proto` sources.
+ */
+fun isGoogleProtoSource(file: FileTreeElement): Boolean {
```
Comment (alexander-yevsyukov) [inline base/build.gradle.kts:137]: Is it possible to do the following in this context:

```kotlin
fun FileTreeElement.isGoogleProtoSource(): Boolean { ... }
```

File: buildSrc/src/main/kotlin/io/spine/gradle/internal/deps.kt [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+    val mockito       = "org.mockito:mockito-core:2.12.0"
+    val hamcrest      = "org.hamcrest:hamcrest-all:1.3"
+    val truth = listOf(
+            "com.google.truth:truth:${Versions.truth}",
+            "com.google.truth.extensions:truth-java8-extension:${Versions.truth}",
+            "com.google.truth.extensions:truth-proto-extension:${Versions.truth}"
+    )
+    @Deprecated("Use Flogger over SLF4J.",
+                replaceWith = ReplaceWith("Deps.runtime.floggerSystemBackend"))
+    @Suppress("DEPRECATION")
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/gradle/internal/deps.kt:221]: I don't understand this suppression. Can you please have a comment?

File: build.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+
+allprojects {
+    apply {
+        plugin("jacoco")
+        plugin("idea")
+        plugin("project-report")
+        from("$rootDir/config/gradle/dependencies.gradle")
+    }
+    version = rootProject.extra["versionToPublish"]!!
+    version = rootProject.extra["versionToPublish"]!!
```
Comment (alexander-yevsyukov) [inline build.gradle.kts:86]: Why do we need this twice?

File: tools/errorprone-checks/build.gradle.kts [hunk 6]
```diff
@@ -18,44 +18,42 @@
...
-    implementation project(':base')
-    implementation project(':plugin-base')
-
-    implementation deps.build.errorProneCore
-    implementation deps.build.errorProneAnnotations
-
-    testImplementation deps.build.errorProneTestHelpers
+    annotationProcessor(Deps.build.autoService.processor)
+    compileOnly(Deps.build.autoService.annotations)
+    implementation(project(":base"))
```
Comment (armiol) [inline tools/errorprone-checks/build.gradle.kts:34]: This one reads better than `"implementation"` in `build.gradle.kt` above, i.e.
```kotlin
        "implementation"(Deps.build.guava)
        "implementation"(Deps.build.checkerAnnotations)
        "implementation"(Deps.build.jsr305Annotations)
```
It is possible to use no-quote syntax there as well?

File: build.gradle.kts [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+    }
+
+    DependencyResolution.defaultRepositories(repositories)
+    dependencies {
+        "errorprone"(Deps.build.errorProneCore)
+        "errorproneJavac"(Deps.build.errorProneJavac)
+        Deps.build.protobuf.forEach { "api"(it) }
+        "api"(Deps.build.flogger)
+        "implementation"(Deps.build.guava)
+        "implementation"(Deps.build.checkerAnnotations)
```
Comment (alexander-yevsyukov) [inline build.gradle.kts:131]: Are there calls that allow to say `implementation` only once, accepting a list of dependencies?

File: tools/plugin-base/build.gradle.kts [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        }
+    }
+}
+
+tasks.processResources.get().dependsOn(prepareProtocConfigVersions)
+
+sourceSets.test {
+    resources.srcDir("$projectDir/src/test/resources")
+}
+
```
Comment (armiol) [inline tools/plugin-base/build.gradle.kts:81]: Please kill these redundant empty lines.

File: build.gradle.kts [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+        "testImplementation"(Deps.test.guavaTestlib)
+        "testImplementation"(Deps.test.junit5Runner)
+        "testImplementation"(Deps.test.junitPioneer)
+        Deps.test.junit5Api.forEach { "testImplementation"(it) }
+        "runtimeOnly"(Deps.runtime.floggerSystemBackend)
+    }
+
+    DependencyResolution.forceConfiguration(configurations)
+    configurations {
+        named("runtime").get().exclude(group = "com.google.protobuf", module = "protobuf-lite")
```
Comment (alexander-yevsyukov) [inline build.gradle.kts:143]: I think this becomes a “reusable code”. Can we extract it somehow?

File: testlib/build.gradle.kts [hunk 10]
```diff
@@ -17,27 +19,18 @@
...
-protobuf {
-    protoc {
-        artifact = deps.build.protoc
-    }
+    Deps.build.protobuf.forEach { api(it) }
+    Deps.test.junit5Api.forEach { api(it) }
+    Deps.test.truth.forEach { api(it) }
+    api(Deps.test.guavaTestlib)
+    api(Deps.test.hamcrest)
+    implementation(project(":base"))
```
Comment (alexander-yevsyukov) [inline testlib/build.gradle.kts:35]: Why is this code uses `api` and `implementation` without quotes, while the code “above” uses similar calls with quotes?

Comment (dmdashenkov) [reply testlib/build.gradle.kts:35]: Resolved. Now it all uses the quote-less (type-safe) accessors. Those were remnants of the early trials.

**Pull request merged**